### PR TITLE
Add semantic segmentation predict

### PR DESCRIPTION
**Pull request by guarin**
## What has changed and why?

* Add predict for EoMT
* Remove `backbone_freeze` for EoMT as it doesn't make sense

## How has it been tested?

* Unit tests
* Manual

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 1]
```diff
@@ -195,6 +295,114 @@ def forward_train(self, x: Tensor) -> tuple[list[Tensor], list[Tensor]]:
             class_logits_per_layer,
         )
 
+    def tile(
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:298]: Moved from `train_model.py`

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 2]
```diff
@@ -195,6 +295,114 @@ def forward_train(self, x: Tensor) -> tuple[list[Tensor], list[Tensor]]:
...
+
+                # Store the crop.
+                crops.append(crop)
+
+                # Store the position of the crop.
+                origins.append((i, start, end, is_tall))
+
+        return crops, origins
+
+    def untile(
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:337]: Moved from `train_model.py`

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 3]
```diff
@@ -195,6 +295,114 @@ def forward_train(self, x: Tensor) -> tuple[list[Tensor], list[Tensor]]:
...
+                logit_sums[image_index][:, :, start:end] += crop_logits[crop_index]
+                logit_counts[image_index][:, :, start:end] += 1
+
+        # Average the logits in the regions of overlap.
+        return [
+            logit_sum / logit_count
+            for logit_sum, logit_count in zip(logit_sums, logit_counts)
+        ]
+
+    def to_per_pixel_logits_semantic(
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:370]: Moved from `train_model.py`. Added adjustment for ignore class logits.

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 4]
```diff
@@ -293,8 +501,3 @@ def load_train_state_dict(self, state_dict: dict[str, Any]) -> None:
                 if name in param_names:
                     new_state_dict[name] = param
         self.load_state_dict(new_state_dict, strict=True)
-
-    def freeze_backbone(self) -> None:
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:297]: Removed as backbone freezing doesn't make sense with EoMT. In the future we could add backbone freezing for the first blocks of the backbone until the query tokens are added. But PEFT might be a better option.

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 5]
```diff
@@ -105,10 +148,67 @@ def __init__(
...
+        x = transforms_functional.normalize(
+            x, mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]
+        )
+        # Resize shorter edge to 518
+        # TODO(Guarin, 07/25): Make this configurable. Save default image size in the
+        # model.
+        x = transforms_functional.resize(x, size=[518])
+        x = x.unsqueeze(0)  # (1, C, H, W)
+
+        logits = self._forward_logits(x)  # (1, C, H, W)
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:189]: Can you update the comment regarding the tensor shape for our future selves?
I think it should be K+1 and H, W are not the same as for the image as this point, right?

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 6]
```diff
@@ -105,10 +148,67 @@ def __init__(
...
+        if self.class_ignore_index is None:
+            # Restrict logits to known classes only.
+            logits = logits[:, :-1]
+        logits = F.interpolate(
+            logits, size=(image_h, image_w), mode="bilinear"
+        )  # (1, C, H, W)
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:195]: Also here K or K+1

File: src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py [hunk 7]
```diff
@@ -195,6 +295,114 @@ def forward_train(self, x: Tensor) -> tuple[list[Tensor], list[Tensor]]:
...
+        # x is a batch of images with shape (B, C, H, W).
+
+        # Tiling.
+        image_sizes = [img.shape[-2:] for img in x]
+        crops_list, origins = self.tile(images=x)
+        crops = torch.stack(crops_list)
+        crop_h, crop_w = crops.shape[-2:]
+
+        # Forward pass.
+        mask_logits_per_layer, class_logits_per_layer = self.forward_train(crops)
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_eomt_semantic_segmentation/task_model.py:393]: For our future selves it would make sense to add a little comment to say this function returns the logits from multiple blocks but we only keep the last one for predictions.

**Review by stegmuel — APPROVED** [review]
Approved with some comments  to help readability.
